src/gpu/generic/sycl/rnn/ref_rnn.cpp

```diff
@@ -0,0 +1,733 @@
+/*******************************************************************************
+* Copyright 2019-2024 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+// General architecture
+//
+// for diff states, we have n_states + 1 as we have n_states diff
+// to propagate to the previous iteration and 1 states to propagate
+// to the previous layer
+// index 0 is dh for cell(t-1, l) to consume
+// index 1 is dc for cell(t-1, l) to consume
+// index 2 is dh for cell(t, l-1) to consume
+// this indexing enables to have the same indexing for states in elemwise
+// function
+// only the cell execution function should be impacted
+
+#include "gpu/generic/sycl/rnn/ref_rnn.hpp"
+#include "common/primitive.hpp"
+#include "common/primitive_desc.hpp"
+
+#include "common/c_types_map.hpp"
+#include "common/gemm_utils.hpp"
+#include "common/type_helpers.hpp"
+#include "gpu/generic/sycl/rnn/rnn_kernels.hpp"
+#include "gpu/intel/gemm/gpu_gemm.hpp"
+
+#include <memory>
+#include "gpu/intel/utils.hpp"
+
+#define DPRINT(fmt, ...) \
```

Comment:
This is actually used for debugging purposes only when the debug mode is set, which is also present for the ocl implementation, I believe it can remain.

Reply:
This adds branching and it can affect the generated code as ressources need to be allocated in case the kernel needs to print something. We can keep the debugging macro but I think it should be enabled only when building in debug mode. `DPRINT` can be defined to be empty when building in release mode.